kernel/kernel-api/src/main/java/io/delta/kernel/internal/clustering/ClusteringUtils.java

```
@@ -0,0 +1,89 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class ClusteringUtils {
+
+  private ClusteringUtils() {
+    // Empty private constructor to prevent instantiation
+  }
+
+  /**
+   * Get the domain metadata for the clustering columns.
+   *
+   * @param logicalClusteringColumns The logical clustering columns.
+   * @param schema The schema of the table.
+   * @return The domain metadata including the clustering columns.
+   */
+  public static DomainMetadata getClusteringDomainMetadata(
+      List<Column> logicalClusteringColumns, StructType schema) {
+    final ClusteringMetadataDomain clusteringMetadataDomain =
+        new ClusteringMetadataDomain(logicalClusteringColumns, schema);
+    return clusteringMetadataDomain.toDomainMetadata();
+  }
+
+  /**
+   * Extract ClusteringColumns from a given snapshot. Return None if the clustering domain metadata
+   * is missing.
+   *
+   * @param snapshot The snapshot instance
+   */
+  public static Optional<List<Column>> getClusteringColumnsOptional(SnapshotImpl snapshot) {
+    return ClusteringMetadataDomain.fromSnapshot(snapshot)
+        .map(ClusteringMetadataDomain::fetchClusteringColumns);
+  }
+
+  /**
+   * Validates the given {@link DataFileStatus} being added as an {@code add} action to the Delta
+   * Log. Specifically, it ensures that file-level statistics are present and that statistics exist
+   * for all clustering columns.
+   *
+   * @param clusteringColumns The list of clustering columns that require statistics.
+   * @param dataFileStatus The {@link DataFileStatus} to validate.
+   */
+  public static void validateDataFileStatus(
+      List<Column> clusteringColumns, DataFileStatus dataFileStatus) {
+    if (!dataFileStatus.getStatistics().isPresent()) {
+      throw DeltaErrors.missingFileStatsForClustering(clusteringColumns, dataFileStatus);
+    }
+
+    DataFileStatistics dataFileStatistics = dataFileStatus.getStatistics().get();
```

Should we throw if missing? `dataFileStatus.getStatistics().orElseThrow(..)`

---

the missing case would be thrown in https://github.com/delta-io/delta/pull/4322/files/08de273bf2ae9e5b78c7c48df1b849e270d48003#diff-1e3d98c63a7251c5f7652ca058b511c52a152b66e5259a811b6296938758924eR71